File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        Stack<Integer> subBelt = new Stack<>(); // 보조 컨테이너 벨트
        int currentBox = 1; // 메인 벨트에서 꺼낼 다음 상자 번호
        
        // 각 상자 순서대로 확인하며 실을 수 있는지 검토
        for (int target : order) {
            // 메인 벨트에서 target 상자를 꺼낼 때까지 subBelt로 이동
            while (currentBox <= target) {
                subBelt.push(currentBox++);
            }

            // 보조 벨트의 최상단이 target과 같다면 실을 수 있음
            if (!subBelt.isEmpty() && subBelt.peek() == target) {
                subBelt.pop();
                answer++;
            } else {
                break; // 실을 수 없는 경우 종료
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.py
def solution(order):
    stack = [] # 보조 컨베이어벨트
    answer = 0 # 실은 상자들
    box = 1
    
    for target in order:
        while box < target and box <= len(order):
            stack.append(box)
            box += 1
        
        # 1. 보조 벨트 top이 o면
        if stack and stack[-1] == target:
            s = stack.pop()
            answer += 1
        # 2. 아니면 기존 벨트에서 상자를 계속 꺼낸다.
        elif box == target:
            answer += 1
            box += 1
        # 3. 기존 벨트에서도 더 이상 꺼낼 수 없고,
        #    보조 벨트 top도 target이 아니면
        #    → break
        else:
            break
        

    return answer